compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -176,16 +176,25 @@ fn build_drop_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, ty: Option<Ty<'tcx>>)
 
     if ty.is_some() {
         // The first argument (index 0), but add 1 for the return value.
-        let dropee_ptr = Place::from(Local::new(1 + 0));
+        let mut dropee_ptr = Place::from(Local::new(1 + 0));
         if tcx.sess.opts.unstable_opts.mir_emit_retag {
             // Function arguments should be retagged, and we make this one raw.
-            body.basic_blocks_mut()[START_BLOCK].statements.insert(
-                0,
-                Statement {
-                    source_info,
-                    kind: StatementKind::Retag(RetagKind::Raw, Box::new(dropee_ptr)),
-                },
+            let reborrow = Rvalue::Ref(
+                tcx.lifetimes.re_erased,
+                BorrowKind::Mut { allow_two_phase_borrow: false },
+                tcx.mk_place_deref(dropee_ptr),
             );
+            let ref_ty = reborrow.ty(body.local_decls(), tcx);
+            dropee_ptr = body.local_decls.push(LocalDecl::new(ref_ty, span)).into();
+            let statements = [
+                StatementKind::Assign(Box::new((dropee_ptr, reborrow))),
+                StatementKind::Retag(RetagKind::FnEntry, Box::new(dropee_ptr)),
+            ];
+            for s in statements {
+                body.basic_blocks_mut()[START_BLOCK]
+                    .statements
+                    .push(Statement { source_info, kind: s });
+            }
```

Comment:
The previous code added things at the *beginning* of the start block, you are moving it to the end. That seems strange. Why did you change that?

I also don't understand why this is so different from the old code -- why is there also an assignment being generated? This needs tons of comments to explain all that happens.

Reply:
Oh, you need the assignment because `Retag` ignores raw pointers. Makes sense but needs a comment.

But the point about inserting at the beginning vs the end remains.

Reply:
There are no statements in the body at this time, so the beginning and end are the same.

I will add comments and more tests once we actually decide if we want this behavior, this PR as is was not meant to be mergable

Reply:
> There are no statements in the body at this time, so the beginning and end are the same.

Then please `assert!` that. I added them at the beginning just to be sure.

Reply:
I've added a comment to this, but the body is created like 10 lines up, so I don't really think it's worth an assert